File: 97-interleaving-string/97-interleaving-string.java
class Solution {
    String s1,s2,s3;
    int [][]dp={};
    int search(int i,int j,int k){
        if(k==s3.length())
            return 1;
        if(dp[i][j]!=-1)
            return dp[i][j];
        if(i<s1.length() && s1.charAt(i)==s3.charAt(k) && search(i+1,j,k+1)==1){
            dp[i][j]=1;
            return dp[i][j];
        }
        if(j<s2.length() && s2.charAt(j)==s3.charAt(k) && search(i,j+1,k+1)==1){
            dp[i][j]=1;
            return dp[i][j];
        }
        dp[i][j]=0;
        return dp[i][j];
    }
    public boolean isInterleave(String s11, String s22, String s33) {
        s1=s11;
        s2=s22;
        s3=s33;
        dp=new int[s1.length()+1][s2.length()+1];
        for(int i=0;i<s1.length()+1;i++)for(int j=0;j<s2.length()+1;j++)dp[i][j]=-1;
        if(s1.length()+s2.length()==s3.length())
            return search(0,0,0)==1;
        return false;
    }
}

File: 97-interleaving-string/97-interleaving-string.cpp
class Solution {
public:
    map<pair<int,int>,int>m;
    
    bool rec(string &s1,string &s2,int i,int j,string &s3){
        if(i==s1.size() && j==s2.size() && i+j==s3.size())
            return m[{i,j}]=true;
        
        if(m.find({i,j})!=m.end())
            return m[{i,j}];
        
        if(i<s1.size() && s1[i]==s3[i+j] && rec(s1,s2,i+1,j,s3))
            return m[{i,j}]=true;
        
        if(j<s2.size() && s2[j]==s3[i+j] && rec(s1,s2,i,j+1,s3))
            return m[{i,j}]=true;
        
        return m[{i,j}]=false;
    }
    
    bool isInterleave(string s1, string s2, string s3) {
        return rec(s1,s2,0,0,s3);
    }
};